source/opt/local_single_block_elim_pass.cpp

```diff
@@ -73,11 +90,21 @@ bool LocalSingleBlockLoadStoreElimPass::LocalSingleBlockLoadStoreElim(
           // variable.
           if (ptrInst->opcode() == SpvOpVariable) {
             // If a previous store to same variable, mark the store
-            // for deletion if not still used.
+            // for deletion if not still used. Don't delete store
+            // for aggregate if debugging; the DebugValue will complicate
+            // scalar replacement.
             auto prev_store = var2store_.find(varId);
             if (prev_store != var2store_.end() &&
-                instructions_to_save.count(prev_store->second) == 0) {
+                instructions_to_save.count(prev_store->second) == 0 &&
+                !(IsArrayOrStruct(varId) &&
+                  get_debug_info_mgr()->IsDebugDeclared(varId))) {
```

Comment:
In my opinion, we should allow delete store for this case. Consider the following example:
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
OpStore %foo %int_13  // will be deleted
OpStore %foo %int_17
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
```
We have to delete `OpStore %foo %int_13`.

The following case looks non-trivial:
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
OpStore %foo %int_13  // will be deleted
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
...
OpStore %foo %int_17
```

IMO, we have to change it to
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
%val = OpExtInst %void %ext DebugValue %dbg_foo %int_13 %empty_expr
...
OpStore %foo %int_17
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
```

What do you think?

Reply:
I think we should just not delete the store for any variable (scalars as well) that have a debug declare.  If DebugDeclare is suppose to match llvm.dbg.declare, the address has to always contains the correct value.  However, the spir-v extension is not as clear as the llvm-ir spec, this might be fine.  

I think that as long as we still remove the loads, this pass will still do what it needs to do.

Reply:
For clarification, do you believe that we should not remove the store even if we replace it with a DebugValue, which is being done here?

More generally, can we allow DebugDeclare and DebugValue to coexist for the same DebugLocalVariable? I think we can temporarily within spirv-opt as long as we ultimately remove the DebugDeclares before outputting the SPIR-V.

Reply:
I don't think we should remove the store.  The DebugValue is not needed, but I don't think it hurts.

> can we allow DebugDeclare and DebugValue to coexist for the same DebugLocalVariable?

The spec is so unclear, I am not sure what we can or cannot do.  I believe it would be okay to have both as long as they give the same information.

>  I think we can temporarily within spirv-opt as long as we ultimately remove the DebugDeclares before outputting the SPIR-V.

If we could guarantee that users always run ssa-rewrite then I could agree with this: temporarily invalid code is fine.  However, our passes are meant to be used in potentially any order.  We have tried to ensure that no pass requires another pass to run in order to be correct.  The only exception so far is merge-return and dead-branch elimination.  In that case an error is emitted (https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/merge_return_pass.cpp#L108).

Reply:
Ah ok, my comment was wrong.
[llvm-dbg-declare](https://llvm.org/docs/SourceLevelDebugging.html#llvm-dbg-declare) explains "It is not control-dependent ... that address is considered to be the true home of the variable across its entire lifetime."

We must not delete the store if the variable is used for `DebugDeclare`.

Reply:
Not sure why you think we should not remove the OpStore.

Do we agree that for a scalar local (function scope) OpVariable with a DebugLocalVariable and a DebugDeclare, it is ok , and actually a good thing (a goal for spirv-opt) to replace all its OpStores with DebugValues, remove the DebugDeclare and the actual OpVariable, but leave the DebugLocalVariable in place?

What I would like to establish is what we want the legalization passes to do and not do in the presense of OpenCL.DebugInfo.100.

Can/should the legalization passes do less optimization in the presense of OpenCL.DebugInfo.100? For example, can/should the passes not eliminate OpStores? I think it is possible to eliminate OpStores as long as we replace them with DebugValues.

Reply:
I think it might be possible to legalize without removing OpStores. I will have to go back to our original legalization shaders and see. I just thought, to be safe, we should try to stick to the original passes as closely as possible but still retain the debug information.

Reply:
I am asking because I am looking at DCE right now which eliminates a lot of OpStores. If you think that is bad, it would be good to know now.

Reply:
> Not sure why you think we should not remove the OpStore.

If you look up the spec for llvm.dbg.declar (which Jaebaek referenced above), it clearly says that if you have an llvm.dbg.declare for a local variable, then that is the address for the variable for its entire lifetime.  If we remove a store, while leaving a debug declare, then that will have been violated.  That is why I don't think we can remove the store.

Reply:
> Do we agree that for a scalar local (function scope) OpVariable with a DebugLocalVariable and a DebugDeclare, it is ok , and actually a good thing (a goal for spirv-opt) to replace all its OpStores with DebugValues, remove the DebugDeclare and the actual OpVariable, but leave the DebugLocalVariable in place?

Yes, I agree.  The ssa-rewriter should be able to do that, and it will get done.  The ssa-rewritter will replace the DebugDeclare with all of the apporpriate DebugValue instruction, and then we can remove the stores.  If the DebugDeclare is still present, we cannot remove the stores.

Reply:
> Can/should the legalization passes do less optimization in the presense of OpenCL.DebugInfo.100? For example, can/should the passes not eliminate OpStores? I think it is possible to eliminate OpStores as long as we replace them with DebugValues.

If the OpVariable is used by a DebugDeclare or a DebugValue then we should not remove the store unless you can also remove the use in the DebugDeclare or DebugValue.  ssa-rewrite if removing the DebugDeclare and DebugValue, so later passes like ADCE can remove the stores.  We can still remove the loads at any time, and do the copy propagation etc.

Reply:
> I think it might be possible to legalize without removing OpStores. I will have to go back to our original legalization shaders and see. I just thought, to be safe, we should try to stick to the original passes as closely as possible but still retain the debug information.

I believe so as well, as long at the loads are removed.  That is what we really wanted from elim-local-single-block and elim-local-single-store anyway.

Reply:
> I am asking because I am looking at DCE right now which eliminates a lot of OpStores. If you think that is bad, it would be good to know now.

As I said before DCE should not remove the store if the OpVariable is used in a DebugDeclare or DebugValue.  The ssa rewriter should remove those uses in cases where we can remove the stores, so I don't think it will hinder optimization.

Reply:
I think it is ok to replace a single OpStore with a DebugValue as long as we eventually replace them all and get rid of the DebugDeclare before outputting the SPIR-V.

I agree with the general philosophy that each pass should be able to stand alone, but elim-local-single-block is not really a stand-alone pass; it was written to be a helper to ssa-rewrite, so I think it is ok to assume that ssa-rewrite will follow.

That said, I don't think its a huge deal if we don't eliminate OpStores in elim-local-single-block. I was just trying to create as little new logic and exceptions and different behavior for DebugInfo as possible in the interest of reducing complexity.

Reply:
OK. Just looked at output from ssa-rewrite. Now I understand your approach better.

ssa-rewrite is adding DebugValue after each Store and removing the DebugDeclare. So essentially ssa-rewrite is preparing the OpStores and OpVariable for deletion.

I can work with this model now that I understand it.